### PR TITLE
start nydus by toml config

### PR DESCRIPTION
**Pull request by raoxiang1996**
Use the toml file to configure the startup parameters of nydus-snapshotter
relates to https://github.com/containerd/nydus-snapshotter/pull/178
Signed-off-by: raoxiang raoxiang1996@gmail.com

### CODE REVIEW

File: cmd/containerd-nydus-grpc/pkg/command/flags.go [hunk 1]
```diff
@@ -230,6 +231,12 @@ func buildFlags(args *Args) []cli.Flag {
 			Usage:       "synchronize `kubernetes.io/dockerconfigjson` secret from kubernetes API server with provided `--kubeconfig-path` (default `$KUBECONFIG` or `~/.kube/config`)",
 			Destination: &args.EnableKubeconfigKeychain,
 		},
+		&cli.StringFlag{
+			Name:        "sanpshotter-config",
```
Comment (imeoer) [inline cmd/containerd-nydus-grpc/pkg/command/flags.go:235]: Better to use `config` directly.

File: config/config.go [hunk 2]
```diff
@@ -69,6 +80,137 @@ type Config struct {
 	RecoverPolicy            string        `toml:"recover_policy"`
 }
 
+type StartupParameterConfig struct {
```
Comment (imeoer) [inline config/config.go:83]: Maybe `StartupFlag`.

File: config/config.go [hunk 3]
```diff
@@ -69,6 +80,137 @@ type Config struct {
...
+	EnableNydusOverlayFS     bool   `toml:"enable_nydus_overlay_fs"`
+	NydusdThreadNum          int    `toml:"nydusd_thread_num"`
+	CleanupOnClose           bool   `toml:"cleanup_on_close"`
+	KubeconfigPath           string `toml:"kubeconfig_path"`
+	EnableKubeconfigKeychain bool   `toml:"enable_kubeconfig_keychain"`
+	RecoverPolicy            string `toml:"recover_policy"`
+}
+
+type SnapshotterConfig struct {
+	StartupParameterCfg StartupParameterConfig `toml:"StartupParameter"`
```
Comment (imeoer) [inline config/config.go:114]: Maybe use this?

```
StartupFlag StartupFlag `toml:"snapshotter"`
```

File: config/config.go [hunk 4]
```diff
@@ -69,6 +80,137 @@ type Config struct {
...
+}
+
+func LoadShotterConfigFile(snapshotterConfigPath string, config *SnapshotterConfig) error {
+	// get nydus-snapshotter configuration from specified path of toml file
+	if snapshotterConfigPath == "" {
+		return errors.New("snapshotter config path cannot be specified")
+	}
+	tree, err := toml.LoadFile(snapshotterConfigPath)
+	if err != nil && !(os.IsNotExist(err)) {
```
Comment (imeoer) [inline config/config.go:123]: Why ignore `os.IsNotExist` ?

File: config/config.go [hunk 5]
```diff
@@ -69,6 +80,137 @@ type Config struct {
...
+		return errors.Wrapf(err, "failed to load snapshotter config file %q", snapshotterConfigPath)
+	}
+	if err := tree.Unmarshal(config); err != nil {
+		return errors.Wrapf(err, "failed to unmarshal snapshotter config file %q", snapshotterConfigPath)
+	}
+	return nil
+}
+
+func SetStartupParameter(startupParameterCfg *StartupParameterConfig, cfg *Config) error {
```
Comment (imeoer) [inline config/config.go:132]: Why not reuse `comman.Validate` ?

Comment (raoxiang1996) [reply config/config.go:132]: If I use `comman.Validate`, I should import package.Circular reference will occur.

File: .idea/workspace.xml [hunk 6]
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (imeoer) [inline .idea/workspace.xml:1]: Why the file is commited?

File: config/config.go (hunk 5 above)
Comment (imeoer) [reply config/config.go]: Maybe move it to a separate file?

**Comment by changweige** [comment]
Can you please paste a demo configuration toml file? That will be more straightforward to check out what can be configured and the configuration hierarchy.

**Comment by raoxiang1996** [comment]
> 

ok，I pasted a demo configuration toml file in doc

**Comment by raoxiang1996** [comment]
> 

ok， I pasted a a demo configuration toml file in doc `configure_nydus.md`

**Comment by raoxiang1996** [comment]
> Can you please paste a demo configuration toml file? That will be more straightforward to check out what can be configured and the configuration hierarchy.

ok， I pasted a a demo configuration toml file in doc `configure_nydus.md`

File: docs/configure_nydus.md [hunk 7]
```diff
@@ -1,8 +1,36 @@
...
+# Nydus-snapshotter startup parameter
+[snapshotter]
+config_path = "/etc/nydusd-config.json" 
+address = "/run/containerd-nydus/containerd-nydus-grpc.sock"
+root_dir  = "root_dir"
+cache_dir  = "/var/lib/nydus/cache"
+nydusd_binary_path  = "/usr/local/bin/nydusd"
+nydus_image_binary_path = "/usr/local/bin/nydus-image"
+log_to_stdout = true
+...
```
Comment (imeoer) [inline docs/configure_nydus.md:18]: Better to list all flags. :)

File: docs/configure_nydus.md [hunk 8]
```diff
@@ -1,8 +1,36 @@
...
+# Nydusd default template
+[nydusd.template]
```
Comment (imeoer) [inline docs/configure_nydus.md:20]: It seems currently this configuration field is not yet supported, can we remove it?

File: docs/configure_nydus.md [hunk 9]
```diff
@@ -1,8 +1,36 @@
...
+
+
+# mirror configuration
+[[mirrors.host."A.com"]]
```
Comment (imeoer) [inline docs/configure_nydus.md:24]: Ditto.

**Comment by imeoer** [comment]
@raoxiang1996 Some conflicts need to be resolved.

File: docs/configure_nydus.md [hunk 10]
```diff
@@ -1,8 +1,42 @@
...
+enable_stargz = true
+disable_cache_manager = false
+enable_nydus_overlay_fs = true
+nydusd_thread_num = 1
+cleanup_on_close = false
+kubeconfig_path = "$HOME/.kube/config"
+enable_kubeconfig_keychain = true
+recover_policy = "restart"
+print_version = true
+```
```
Comment (changweige) [inline docs/configure_nydus.md:39]: It's better that we can classify the above configuration items into several groups. Then we will have a configurations hierarchy. Don't worry, I will try to come up with a possible rule to classify them. Simply flattening them is not that clear.

File: cmd/containerd-nydus-grpc/pkg/command/flags.go [hunk 11]
```diff
@@ -7,61 +7,49 @@
...
+	EnableStargz             bool   `toml:"enable_stargz"`
+	DisableCacheManager      bool   `toml:"disable_cache_manager"`
+	LogToStdout              bool   `toml:"log_to_stdout"`
+	EnableNydusOverlayFS     bool   `toml:"enable_nydus_overlay_fs"`
+	NydusdThreadNum          int    `toml:"nydusd_thread_num"`
+	CleanupOnClose           bool   `toml:"cleanup_on_close"`
+	KubeconfigPath           string `toml:"kubeconfig_path"`
+	EnableKubeconfigKeychain bool   `toml:"enable_kubeconfig_keychain"`
+	RecoverPolicy            string `toml:"recover_policy"`
+	PrintVersion             bool   `toml:"print_version"`
```
Comment (changweige) [inline cmd/containerd-nydus-grpc/pkg/command/flags.go:52]: `PrintVersion` should only come from CLI flags. It intends to print build-time info by invoking the command. Configurations in toml should only apply to those parameters for long-time running.

File: cmd/containerd-nydus-grpc/pkg/command/flags.go [hunk 12]
```diff
@@ -7,61 +7,49 @@
...
+	ConfigPath               string `toml:"config_path"`
+	RootDir                  string `toml:"root_dir"`
+	CacheDir                 string `toml:"cache_dir"`
+	GCPeriod                 string `toml:"gc_period"`
+	ValidateSignature        bool   `toml:"validate_signature"`
+	PublicKeyFile            string `toml:"public_key_file"`
+	ConvertVpcRegistry       bool   `toml:"convert_vpc_registry"`
+	NydusdBinaryPath         string `toml:"nydusd_binary_path"`
+	NydusImageBinaryPath     string `toml:"nydus_image_binary_path"`
+	SharedDaemon             bool   `toml:"shared_daemon"`
```
Comment (changweige) [inline cmd/containerd-nydus-grpc/pkg/command/flags.go:37]: `SharedDaemon` should already be deprecated. It is kept in CLI flags for back compatibility. So we can safely reject it into nydus-snapshotter's configuration file

File: config/config.go [hunk 13]
```diff
@@ -69,6 +82,108 @@ type Config struct {
 	RecoverPolicy            string        `toml:"recover_policy"`
 }
 
+type SnapshotterConfig struct {
+	StartupFlag command.Args `toml:"snapshotter"`
+}
+
+func LoadShotterConfigFile(snapshotterConfigPath string, config *SnapshotterConfig) error {
```
Comment (changweige) [inline config/config.go:89]: LoadSnapshotterConfigFile ?

Comment (changweige) [reply config/config.go:89]: Can we return snapshotter's config from the function rather using output parameters?

**Comment by changweige** [comment]
image
Can you remove the merge commit? Try to rebase your branch to current main HEAD

**Pull request closed without merging**